community/io/src/test/java/org/neo4j/graphdb/mockfs/DelegatingFileSystem.java

```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        delegate.close();
+    }
+
+    @Override
+    public boolean isOpen()
+    {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public boolean isReadOnly()
+    {
+        return delegate.isReadOnly();
+    }
+
+    @Override
+    public String getSeparator()
+    {
+        return delegate.getSeparator();
+    }
+
+    @Override
+    public Iterable<Path> getRootDirectories()
+    {
+        final Iterable<Path> rootDirectories = delegate.getRootDirectories();
+        return new Iterable<Path>()
+        {
+            @Override
+            public Iterator<Path> iterator()
+            {
+                final Iterator<Path> iterator = rootDirectories.iterator();
+                return new Iterator<Path>()
+                {
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return iterator.hasNext();
+                    }
+
+                    @Override
+                    public Path next()
+                    {
+                        return createDelegate( iterator.next() );
+                    }
+
+                    @Override
+                    public void remove()
+                    {
+                        iterator.remove();
+                    }
+                };
+            }
+        };
+    }
+
+    @Override
+    public Iterable<FileStore> getFileStores()
+    {
+        return delegate.getFileStores();
+    }
+
+    @Override
+    public Set<String> supportedFileAttributeViews()
+    {
+        return delegate.supportedFileAttributeViews();
+    }
+
+    @Override
+    public Path getPath( String first, String... more )
+    {
+        return createDelegate( delegate.getPath( first, more ) );
+    }
+
+    @Override
+    public PathMatcher getPathMatcher( String syntaxAndPattern )
+    {
+        final PathMatcher matcher = delegate.getPathMatcher( syntaxAndPattern );
+        return new PathMatcher()
+        {
+            @Override
+            public boolean matches( Path path )
+            {
+                return matcher.matches( DelegatingPath.getDelegate( path ) );
+            }
+        };
+    }
+
+    @Override
+    public UserPrincipalLookupService getUserPrincipalLookupService()
+    {
+        return delegate.getUserPrincipalLookupService();
+    }
+
+    @Override
+    public WatchService newWatchService() throws IOException
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    protected Path createDelegate( Path path )
+    {
+        return new WrappedPath( path );
+    }
+
+    protected DelegatingFileSystemProvider createDelegate( FileSystemProvider provider )
+    {
+        return new WrappedProvider( provider, this );
+    }
+
+    public class WrappedPath extends DelegatingPath
+    {
+        WrappedPath( Path delegate )
+        {
+            super( delegate, fileSystem );
```

Comment:
You can use `DelegatingFileSystem.this` and drop the `fileSystem` field. Also, any reason `WrappedPath` isn't static when `WrappedProvider` is?